addons/craft/craft.lua

```diff
@@ -0,0 +1,738 @@
+        local id, name = next(search, nil)
+        if id then
+            notice("Setting auto food to %s":format(name.en))
+            food = name.en
+        else
+            return "invalid food %s":format(item)
+        end
+    end
+end
+
+local function handle_put(ingredient, bag)
+    if bag then
+        bag = bag:lower()
+        if not bags[bag] then
+            return "unknown bag %s":format(bag)
+        end
+    end
+    local search = res.items:name(ingredient)
+    local id, name = next(search, nil)
+    if id then
+        local msg = nil
+        local args = {
+            ['id'] = id,
+            ['bag'] = bag,
+            ['name'] = name.en,
+        }
+        local item = {put, args}
+        if bag then
+            msg = "%s %s":format(ingredient, bag)
+        else
+            msg = ingredient
+        end
+        notice("Adding a put %s command to the queue":format(msg))
+        queue:push(item)
+        process_queue()
+    else
+        return "unknown item %s":format(ingredient)
+    end
+end
+
+local function display_crafting_packet(id, data)
+    if id == 0x096 and display then
+        local p = packets.parse('outgoing', data)
+        log(p)
+    end
+end
+
+local function handle_display()
+    if display then
+        notice("Disabling display.")
+        display = false
+    else
+        notice("Enabling display.")
+        display = true
+    end
+end
+
+local function handle_support()
+    if support then
+        notice("Disabling support.")
+        support = false
+    else
+        notice("Enabling support.")
+        support = true
+    end
+end
+
+local function handle_find(query, details)
+    local query = query:lower()
+    notice("Searching for recipes containing %s":format(query))
+    for name, recipe in pairs(recipes) do
+        if string.find(name:lower(), query) then
+            notice("Found recipe - \"%s\"":format(name))
+            if details then
+                notice(" %s":format(recipe['crystal']))
+                for _, ingredient in pairs(recipe['ingredients']) do
+                    notice("  %s":format(ingredient))
+                end
+            end
+        end
+    end
+end
+
+handlers['clear'] = handle_clear
+handlers['repeat'] = handle_repeat
+handlers['r'] = handle_repeat
+handlers['delay'] = handle_delay
+handlers['pause'] = handle_pause
+handlers['resume'] = handle_resume
+handlers['make'] = handle_make
+handlers['m'] = handle_make
+handlers['display'] = handle_display
+handlers['put'] = handle_put
+handlers['food'] = handle_food
+handlers['status'] = handle_status
+handlers['help'] = handle_help
+handlers['jiggle'] = handle_jiggle
+handlers['support'] = handle_support
+handlers['find'] = handle_find
+
+local function handle_command(...)
+    local cmd  = (...) and (...):lower()
+    local args = {select(2, ...)}
+    if handlers[cmd] then
+        local msg = handlers[cmd](unpack(args))
+        if msg then
+            error(msg)
```

Comment:
This is part of the design I opted for because it's simple and because I like the appearance of the code.  Functions can return something, or a string.  If it's a string it indicates something went wrong.